components/motor/dimensionengineering/sabertooth.go

```diff
@@ -0,0 +1,539 @@
+	}
+	if cfg.MotorChannel < 1 || cfg.MotorChannel > 2 {
+		return utils.NewConfigValidationError(path, errors.New("invalid channel, acceptable values are 1 and 2"))
+	}
+	if cfg.SerialAddress < 128 || cfg.SerialAddress > 135 {
+		return utils.NewConfigValidationError(path, errors.New("invalid address, acceptable values are 128 thru 135"))
+	}
+	if cfg.BaudRate != 2400 && cfg.BaudRate != 9600 && cfg.BaudRate != 19200 && cfg.BaudRate != 38400 && cfg.BaudRate != 115200 {
+		return utils.NewConfigValidationError(path, errors.New("invalid baud_rate, acceptable values are 2400, 9600, 19200, 38400, 115200"))
+	}
+	if cfg.MinPowerPct < 0.0 || cfg.MinPowerPct > cfg.MaxPowerPct {
+		return utils.NewConfigValidationError(path, errors.New("invalid min_power_pct, acceptable values are 0 to max_power_pct"))
+	}
+	if cfg.MaxPowerPct > 1.0 {
+		return utils.NewConfigValidationError(path, errors.New("invalid max_power_pct, acceptable values are min_power_pct to 100.0"))
+	}
+
+	return nil
+}
+
+func init() {
+	controllers = make(map[string]*controller)
+
+	_motor := registry.Component{
+		Constructor: func(ctx context.Context, _ registry.Dependencies, config config.Component, logger golog.Logger) (interface{}, error) {
+			conf, ok := config.ConvertedAttributes.(*Config)
+			if !ok {
+				return nil, rdkutils.NewUnexpectedTypeError(conf, config.ConvertedAttributes)
+			}
+			return NewMotor(ctx, conf, logger)
+		},
+	}
+	registry.RegisterComponent(motor.Subtype, modelName, _motor)
+
+	config.RegisterComponentAttributeMapConverter(
+		motor.SubtypeName,
+		modelName,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Config
+			decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{TagName: "json", Squash: true, Result: &conf})
+			if err != nil {
+				return nil, err
+			}
+			if err := decoder.Decode(attributes); err != nil {
+				return nil, err
+			}
+			return &conf, nil
+		},
+		&Config{},
+	)
+}
+
+func newController(c *Config, logger golog.Logger) (*controller, error) {
+	if c.SerialAddress < 128 || c.SerialAddress > 135 {
+		return nil, errors.New("address is out of range. valid addresses are 127 to 135")
+	}
+	ctrl := new(controller)
+	ctrl.activeAxes = make(map[int]bool)
+	ctrl.serialDevice = c.SerialPath
+	ctrl.logger = logger
+	ctrl.address = c.SerialAddress
+
+	if c.TestChan != nil {
+		ctrl.testChan = c.TestChan
+	} else {
+		serialOptions := serial.OpenOptions{
+			PortName:          c.SerialPath,
+			BaudRate:          uint(c.BaudRate),
+			DataBits:          8,
+			StopBits:          1,
+			MinimumReadSize:   1,
+			RTSCTSFlowControl: true,
+		}
+
+		port, err := serial.Open(serialOptions)
+		if err != nil {
+			return nil, err
+		}
+		ctrl.port = port
+	}
+
+	ctrl.activeAxes[1] = false
+	ctrl.activeAxes[2] = false
+
+	return ctrl, nil
+}
+
+// NewMotor returns a Sabertooth driven motor.
+func NewMotor(ctx context.Context, c *Config, logger golog.Logger) (motor.LocalMotor, error) {
+	globalMu.Lock()
```

Comment:
defer unlock please. you have a return in the if-statement that would leave this mutex locked.